feat(curriculum): Create TypeScript Quiz (#56460)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/quiz-typescript/66f1ae758b77cfc3e4da6151.md

#### --text--

What is Typescript?

#### --distractors--

A framework used for system programming.

---

A database management tool.

---

A programming language used for styling websites.

#### --answer--

A superset of JavaScript that adds static typing.

### --question--

#### --text--

Why should TypeScript be used in a project?

#### --distractors--

> Eliminates need for testing in applications.

Reviewer: ```suggestion
It eliminates the need for testing in applications.
```

---

It is faster than all other programming languages.

---

It makes styling websites easier.

#### --answer--

> Improves maintainability by adding type definitions to JavaScript.

Reviewer: ```suggestion
It improves maintainability by adding type definitions to JavaScript.
```

### --question--

#### --text--

What language is TypeScript compiled down to?

#### --distractors--

Java

---

Python

---

C++

#### --answer--

> JavaScript

Reviewer: ```suggestion

JavaScript
```

### --question--

#### --text--

>     function sum(a: number, b: number): number {
> 	    return a + b;
> 	}
> 	console.log(sum(2, 3));
> 

Reviewer 2: code blocks need to be fenced, and here you also need words for your question, a code snippet is not enough to understand what it is being asked
````suggestion
Question using words goes here

```ts
    function sum(a: number, b: number): number {
	    return a + b;
	}
	console.log(sum(2, 3));

```
````

Reviewer 3: The code needs to be formatted as well.

Author: Fixed the formatting. Thanks for pointing it out. @lasjorg

}
console.log(sum(2, 3));
```

#### --distractors--

>  `2 + 3`

Reviewer: ```suggestion
`2 + 3`
```

---

`undefined`

---

`Error: Type 'number' is not assignable to type 'string'`

#### --answer--

`5`

### --question--

#### --text--

What is the primary purpose of an interface in TypeScript?

#### --distractors--

To compile TypeScript code into JavaScript.

---

To create a new data type.

---

> To create a new JavaScript Object

Reviewer: ```suggestion
To create a new JavaScript Object.
```

To create a new JavaScript Object

#### --answer--

To define the structure of an object.

### --question--

#### --text--

> Which of the following is NOT a basic type in TypeScript?

Reviewer 3: ```suggestion
Which of the following is NOT a primitive type in TypeScript?
```
Just a suggestion, but the word primitive seems more accurate.

#### --distractors--

string

---

boolean

---

number

#### --answer--

array

### --question--

#### --text--

Which of the following correctly defines a literal type in TypeScript?

#### --distractors--

`let y: string;`

---

`let z: boolean;`

---

`let a: number[];`

#### --answer--

`let x: 5;`

### --question--

#### --text--

What is the primary purpose of the `type` keyword in TypeScript?

#### --distractors--

> To declare variables

Reviewer 3: ```suggestion
To declare variables.
```

---

> To create classes

Reviewer 3: ```suggestion
To create classes.
```

---

> To implement interfaces

Reviewer 3: ```suggestion
To implement interfaces.
```

#### --answer--

> To define custom types

Reviewer 3: ```suggestion
To define custom types.
```

### --question--

#### --text--

Which of the following correctly defines a union type?

#### --distractors--

`type Result = Pass && Fail;`

---

`type Result = Pass & Fail;`

---

`type Result = Pass + Fail;`

#### --answer--

`type Result = Pass | Fail;`

### --question--

#### --text--

Which of the following capabilities is unique to `interface` and CANNOT be achieved using `type` in TypeScript?

#### --distractors--

> Defining object shapes

Reviewer 3: ```suggestion
Defining object shapes.
```

---

> Creating union types

Reviewer 3: ```suggestion
Creating union types.
```

---

> Creating intersection types

Reviewer 3: ```suggestion
Creating intersection types.
```

#### --answer--

> Implementing a class

Reviewer 3: Not exactly sure what is meant by, implementing a class?
If it is referring to the use of `extends` could we be more explicit about it?

Author: Thanks for pointing it out. I should've phrased that better.
What I meant was that classes can implement interfaces or Interfaces can be used to declare class implementations.

Thanks again

> Declaring class implementations

Reviewer 3: ```suggestion
Declaring class implementations.
```

### --question--

#### --text--

What is the primary purpose of using generics in TypeScript?

#### --distractors--

To implement inheritance between classes.

---

To define structure of an object.

---

> To create a new data type
> 

Reviewer: ```suggestion
To create a new data type.
```

#### --answer--

To write reusable code that can work with multiple types.

### --question--

#### --text--

Which of the following is a **proper example** of a generic function in TypeScript?

#### --distractors--

> 	function add(a: number, b: number): number { return a + b; }

Reviewer: ```suggestion
`function add(a: number, b: number): number { return a + b; }`
```

---

> 	function multiply(a: number, b: number): { return a * b; }

Reviewer: ```suggestion
`function multiply(a: number, b: number): { return a * b; }`
```

---

> 	function getValue<T>(value: any): any { return value; }

Reviewer: ```suggestion
`function getValue<T>(value: any): any { return value; }`
```

#### --answer--

> 	function identity<T>(value: T): T { return value; }

Reviewer: ```suggestion
`function identity<T>(value: T): T { return value; }`
```

### --question--

#### --text--

What is type narrowing in TypeScript?

#### --distractors--

> 
> Reducing the size of compiled JavaScript

Reviewer: ```suggestion
Reducing the size of compiled JavaScript
```

> Reducing the size of compiled JavaScript

Reviewer 3: ```suggestion
Reducing the size of compiled JavaScript.
```

---

> Limiting the number of types in a union

Reviewer 3: ```suggestion
Limiting the number of types in a union.
```

Limiting the number of types in a union

---

> Removing unused types from a TypeScript project.

Reviewer: ```suggestion
Removing unused types from a TypeScript project
```

> Removing unused types from a TypeScript project

Reviewer 3: ```suggestion
Removing unused types from a TypeScript project.
```

#### --answer--

> To write Code that narrows down a union type to a more specific type

Reviewer: ```suggestion
To write code that narrows down a union type to a more specific type
```

> To write code that narrows down a union type to a more specific type

Reviewer 3: ```suggestion
To write code that narrows down a union type to a more specific type.
```

### --question--

#### --text--

Which file is typically used to configure TypeScript compiler options?

#### --distractors--

`package.json`

---

> `package.json`

Reviewer: These should not be the same.

Author: Fixed it. Thanks for pointing it out.

---

`typescript.config.js`

#### --answer--

`tsconfig.json`

### --question--

#### --text--

What does the `target` option in the `tsconfig.json` file specify?

#### --distractors--

The version of TypeScript to use during compilation.

---

The target runtime environment (e.g., Node.js or browser) for the TypeScript code.

---

The specific module system to use for imports and exports.

#### --answer--

> 
> The ECMAScript version to which the TypeScript compiler will transpile the code.

Reviewer: ```suggestion
The ECMAScript version to which the TypeScript compiler will transpile the code.
```

#### --text--

What does the `noImplicitReturns` option in the `tsconfig.json` file enforce?

#### --distractors--

It forces all functions to return a `number`.

---

It requires that all functions return either `true` or `false`.

---

It prevents using the `any` type in TypeScript code.

#### --answer--

> It warns if a function does not return a value.

Reviewer 3: ```suggestion
Checks all code paths in a function to ensure they return a value.
```
Might be a worthwhile distinction to add.

All functions implicitly return `undefined` and you can implicitly/explicitly have `undefined` as the return type, and Typescript will let the function "bottom out" without an explicit return.

---

Works
```
function isTrue(test: boolean) {
  if (test) {
    console.log("true");
  } else {
    return undefined;
  }
  console.log("Bottom");
}
isTrue(true);
```
Doesn't work
```
function isTrue(test: boolean) {
  if (test) {
    console.log("true");
  } else {
    return false;
  }
  console.log("Bottom");
}
isTrue(true);
```

Author: Got it. Learnt something new today.
Added the suggested changes

### --question--

#### --text--

Which TypeScript option prevents implicit `any` types?

#### --distractors--

`noImplicitReturns`

---

`alwaysStrict`

---

`strictNullChecks`

#### --answer--

`noImplicitAny`

### --question--

#### --text--

What is the purpose of the `never` type in TypeScript?

#### --distractors--

To represent null or undefined.

---

To represent any possible value.

---

To represent a type that can be anything except null.

#### --answer--

To represent the type of values that never occur.

### --question--

#### --text--

> What will be the output of the following TypeScript code?
> 
>     enum Color {
>     	 Red,
>     	 Blue
>     }
>     console.log(Color.Red, Color.Blue);
> 
> 

Reviewer 2: ```suggestion
What will be the output of the following TypeScript code?

```ts
    enum Color {
    	 Red,
    	 Blue
    }
    console.log(Color.Red, Color.Blue);
```

```

}
console.log(Color.Red, Color.Blue);

> ```

Reviewer: ````suggestion
```

````

> ```ts
>     enum Color {
>     	 Red,
>     	 Blue
>     }
>     console.log(Color.Red, Color.Blue);

Reviewer 3: ````suggestion
```ts
enum Color {
  Red,
  Blue,
}
console.log(Color.Red, Color.Blue);
```
````

The code block isn't closed, and the formatting is off.

`Red Blue`

---

`1, 2`

---

`Red 2`

#### --answer--

`0 1`

### --question--

#### --text--

How can you specify that a function parameter is optional in TypeScript?

#### --distractors--

> By using the `optional` keyword

Reviewer 3: ```suggestion
By using the `optional` keyword.
```

---

> By setting a default value

Reviewer 3: ```suggestion
By setting a default value.
```

By using the `any` type

#### --answer--

> By adding a `?` after the parameter name
> 
> 
> 

Reviewer 2: ```suggestion
By adding a `?` after the parameter name

```

Author: Got it! I've made the suggested changes. Thanks

> By using the `any` type

Reviewer 3: ```suggestion
By using the `any` type.
```

> By adding a `?` after the parameter name

Reviewer: ```suggestion
By adding a `?` after the parameter name

```
